### PR TITLE
jobs: introduce job_type column to system.jobs

### PR DESCRIPTION
**Pull request by jayshrivastava**
This change introduces an upgrade to add the `job_type` column to
`system.jobs` as a nullable string. This upgrade also creates an
index for the column and backfills it.

With this change, new jobs being inserted into the jobs table by
the jobs registry will populate the type column.

To make migrations easier, this change adds a new builtin,
`crdb_internal.job_payload_type`, which can be used to determine
the type of a jobs payload.

Release note: None
Epic: https://cockroachlabs.atlassian.net/browse/CRDB-19709
Informs: https://github.com/cockroachdb/cockroach/issues/92261

### CODE REVIEW

File: pkg/upgrade/upgrades/alter_jobs_add_job_type.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+)
+
+const addTypeColumnStmt = `
+ALTER table system.jobs
+ADD COLUMN IF NOT EXISTS type STRING
```
Comment (dt) [inline pkg/upgrade/upgrades/alter_jobs_add_job_type.go:25]: drive-by: what would you think about calling it `job_type` since type is a reserved word in some/many places?

that's what the revamp RFC proposed here: https://github.com/cockroachdb/cockroach/pull/82638/files#diff-d65357c15cffdb009c1b3fd0f6456265d25b94758f326207c531b8f734b59164R126

File: pkg/upgrade/upgrades/schema_changes.go [hunk 2]
```diff
@@ -172,6 +173,12 @@ func ensureProtoMessagesAreEqual(expected, found protoutil.Message) error {
 	if bytes.Equal(expectedBytes, foundBytes) {
 		return nil
 	}
+	fmt.Println("expected")
+
+	fmt.Println(expected)
+	fmt.Println("found")
+
+	fmt.Println(found)
```
Comment (andyyang890) [inline pkg/upgrade/upgrades/schema_changes.go:181]: nit: delete these print statements

File: pkg/sql/catalog/systemschema/system.go [hunk 3]
```diff
@@ -1410,6 +1413,19 @@ var (
...
 			},
+			descpb.IndexDescriptor{
+				Name:                "jobs_type_idx",
+				ID:                  5,
+				Unique:              false,
+				KeyColumnNames:      []string{"type"},
+				KeyColumnDirections: []catpb.IndexColumn_Direction{catpb.IndexColumn_ASC},
+				KeyColumnIDs:        []descpb.ColumnID{12},
+				StoreColumnNames:    []string{},
+				StoreColumnIDs:      []descpb.ColumnID{},
```
Comment (andyyang890) [inline pkg/sql/catalog/systemschema/system.go:1424]: nit: I don't think you need to define these if they're empty

File: pkg/upgrade/upgrades/alter_jobs_add_job_type.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	op = operation{
+		name:           "add-jobs-type-col-idx",
+		schemaList:     []string{"jobs_type_idx"},
+		query:          addTypeColumnIdxStmt,
+		schemaExistsFn: hasIndex,
+	}
+	if err := migrateTable(ctx, cs, d, op, keys.JobsTableID, systemschema.JobsTable); err != nil {
+		return err
+	}
```
Comment (andyyang890) [inline pkg/upgrade/upgrades/alter_jobs_add_job_type.go:55]: You could put these into a for loop so that you only need to write the code calling `migrateTable` once.

File: pkg/upgrade/upgrades/alter_jobs_add_job_type.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+UPDATE system.jobs
+SET type = crdb_internal.payload_type(payload)
+WHERE type IS NULL
+`
+
+func backfillSystemJobsTypeColumn(
+	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps,
+) error {
+
+	_, err := d.InternalExecutor.Exec(ctx, "backfill-jobs-type-column", nil /* txn */, backfillTypeColumnStmt, username.RootUser)
```
Comment (andyyang890) [inline pkg/upgrade/upgrades/alter_jobs_add_job_type.go:70]:
```suggestion
	ie := d.InternalExecutorFactory.MakeInternalExecutorWithoutTxn()
	_, err := ie.Exec(ctx, "backfill-jobs-type-column", nil /* txn */, backfillTypeColumnStmt, username.RootUser)
```

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 12 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @dt, and @jayshrivastava)

___
*[`pkg/clusterversion/cockroach_versions.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgyStc1WdWviqU4sgE:-NIgyStc1WdWviqU4sgF:bhezms5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/clusterversion/cockroach_versions.go#L349)):*
> ```Go
> 	// job rows created prior to V23_1AddTypeColumnToJobsTable.
> 	V23_1BackfillTypeColumnInJobsTable
> ```

Just leaving this as I read the code: I'm not sure why we need to have 2 versions; 
can't this be done with alter table/update records?
___
*[`pkg/jobs/registry.go` line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIguSeD6mWpZb9OPbfv:-NIguSeD6mWpZb9OPbfw:b-p115ly) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L383)):*
> ```Go
> 	instanceID := r.ID()
> 	var numColumns = 7
> 	columns := make([]string, 7, 8)
> ```

it's a bit sad that we have to allocate dynamic slice whereas before we had static array.
You could have 
````
columns := [8]{seven elements are initialized only}
numColums = 7

.. if version Okay {
  numColumns = 8
  columns[7] = job type
}
```
````

___
*[`pkg/jobs/registry.go` line 423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgyvL66OJmQCLtPCPJ:-NIgyvL66OJmQCLtPCPK:b-mxxqr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L423)):*
> ```Go
> 		columns = append(columns, `type`)
> 		valueFns[`type`] = func(rec *Record) interface{} {
> 			return (&jobspb.Payload{Details: jobspb.WrapPayloadDetails(rec.Details)}).Type().String()
> ```

Have you considered having job_type column be non-null, with `DEFAULT crdb_internal.payload_type(payload)`?  Basically, have this column computed?

I understand there might be some concern that calling this funciton is "more work"  (i.e. we'd need to unmarshal proto we just marshaled) -- and maybe it's a real concern -- but I think it probably won't matter much and might make the code simpler/smaller?

@ajwerner , wdyt?
___
*[`pkg/jobs/registry.go` line 517 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgr8Pm2lMDE0PYJFlf:-NIgr8Pm2lMDE0PYJFlg:b-y8ohqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L517)):*
> ```Go
> 	if _, err = j.registry.ex.Exec(ctx, "job-row-insert", txn, `
> INSERT INTO system.jobs (id, status, payload, progress, claim_session_id, claim_instance_id, type)
> VALUES ($1, $2, $3, $4, $5, $6, $7)`, jobID, StatusRunning, payloadBytes, progressBytes, s.ID().UnsafeBytes(), r.ID(), jobType.String(),
> ```

wdyt about:

```
cols = []string{id, status, payload, progress, claim_session_id, claim_instance_id}
vals = []interface{}{jobID, StatusRunnnng, payloadBytes, progressBytes, s.ID().UnsafeBytes(), r.ID()}
placeholders := func() string {
  var p strings.Builder
  for i := range vals {
		if i > 0 {
			p.WriteByte(',')
		}
		p.WriteByte('$')
                p.WriteString(strconv.Itoa(i+1))
	}
	return p.String()
}
if r.settings.Version.IsActive(ctx, clusterversion.V23_1AddTypeColumnToJobsTable) {
   cols = append(cols, "type")
   vals = append(vals, jobType.String())
}
insertQuery = fmt.Sprintf(`INSERT INTO system.jobs (%s) VALUES (%s)`, strings.Join(cols, ","),  placeholders())

// Execute query ...
```
___
*[`pkg/sql/sem/builtins/builtins.go` line 4033 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgyLoPDujiy8AOa28d:-NIgyLoPDujiy8AOa28e:b-9mh3xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/sql/sem/builtins/builtins.go#L4033)):*
> ```Go
> 		}()...),
> 
> 	"crdb_internal.payload_type": makeBuiltin(
> ```

nit: perhaps be explicit and name the function job_payload_type()?
___
*[`pkg/sql/sem/builtins/builtins.go` line 4044 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgxvYh08qUB0zYiuNI:-NIgxvYh08qUB0zYiuNJ:b-vx3p0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/sql/sem/builtins/builtins.go#L4044)):*
> ```Go
> 					return nil, pgerror.Wrap(err, pgcode.InvalidParameterValue, "invalid protocol message")
> 				}
> 				typ := msg.(*jobspb.Payload).Type().String()
> ```

you might be better off not using protoreflect package at all (which really exists so that you can create
protocol message of appropriate type, given type name).  You don't need it, because you know exactly the type of the message you expect.  So, something like this:

```
var msg jobspb.Payload
if  if err := protoutil.Unmarshal(data, &msg); err != nil {
   return nil, err...
}
return tree.NewDString(msg.Type().String()), nil
```
___
*[`pkg/sql/sem/builtins/builtins.go` line 4051 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgw06R8-FbERocxoJL:-NIgw06R8-FbERocxoJM:b8zui2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/sql/sem/builtins/builtins.go#L4051)):*
> ```Go
> 				{
> 					Info:       info,
> 					Volatility: volatility,
> ```

honestly, I'd just inline info, volatility, and return type here.  It's
pretty idiomatic in this file; also, why have variables when these are not
variables?
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIcfpQFIYvbyAi8tYYL-r1-25:-NIgvLeP26i25YrdpQYy:b542bjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type.go#L25)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

drive-by: what would you think about calling it `job_type` since type is a reserved word in some/many places?

that's what the revamp RFC proposed here: https://github.com/cockroachdb/cockroach/pull/82638/files#diff-d65357c15cffdb009c1b3fd0f6456265d25b94758f326207c531b8f734b59164R126
</blockquote></details>

Seconded.
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIglnmMA3Ofx_xVzXv4:-NIglnmMA3Ofx_xVzXv5:b-2gwy4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go#L93)):*
> ```Go
> 		record := jobs.Record{
> 			Details:  jobspb.JobDetailsForEveryJobType[typ],
> 			Progress: jobspb.ImportProgress{},
> ```

is this safe to do?  Can this job start running (and likely crash)?
Or is it safe to do because the jobs system is not running yet? 
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go` line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIglJ9VB3fGES0P8I_e:-NIglJ9VB3fGES0P8I_f:b-lqikf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go#L154)):*
> ```Go
> }
> 
> func getDeprecatedJobsTableDescriptor() *descpb.TableDescriptor {
> ```

nit: it's not really deprecated, but I see what you mean.  Maybe call it `getJobsTableDescriptorPrioToV222_....` -- quite a bit of a mouth full, but very descriptive.
Up to you.

File: pkg/upgrade/upgrades/alter_jobs_add_job_type.go (hunk 1 above)
Comment (jayshrivastava) [reply pkg/upgrade/upgrades/alter_jobs_add_job_type.go]: Sgtm 👍🏽

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andyyang890, @dt, and @miretskiy)

___
*[`pkg/clusterversion/cockroach_versions.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgyStc1WdWviqU4sgE:-NIslMcP1HKv_R7ew6EJ:b-sodztz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/clusterversion/cockroach_versions.go#L349)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Just leaving this as I read the code: I'm not sure why we need to have 2 versions; 
can't this be done with alter table/update records?

</blockquote></details>

Sure. We can do everything in one upgrade.
___
*[`pkg/jobs/registry.go` line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIguSeD6mWpZb9OPbfv:-NIsimFM6Xqu4WV9hLWN:bcpwe3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L383)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

it's a bit sad that we have to allocate dynamic slice whereas before we had static array.
You could have 
````
columns := [8]{seven elements are initialized only}
numColums = 7

.. if version Okay {
  numColumns = 8
  columns[7] = job type
}
```
````

</blockquote></details>

Done. How do you feel about defaulting to `numColumns=8` and subtracting 1 if the version is old? The most common path is the one where the version is above the one with my upgrade. In the most common path, we can avoid the if branch.
___
*[`pkg/jobs/registry.go` line 423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgyvL66OJmQCLtPCPJ:-NIh9ag_4GUPYcJg6B9Q:b-p9yuys) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L423)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Have you considered having job_type column be non-null, with `DEFAULT crdb_internal.payload_type(payload)`?  Basically, have this column computed?

I understand there might be some concern that calling this funciton is "more work"  (i.e. we'd need to unmarshal proto we just marshaled) -- and maybe it's a real concern -- but I think it probably won't matter much and might make the code simpler/smaller?

@ajwerner , wdyt?
</blockquote></details>

I considered both options and thought my way was better. I'm open to changing it though.

In `createJobsInBatchWithTxn`, we do `(&jobspb.Payload{Details: jobspb.WrapPayloadDetails(rec.Details)}).Type().String()`, which just allocates some nested structs. In `CreateJobWithTxn`, we do even less work - the job type is in scope and we insert `jobType.String()` directly.

If we use, `DEFAULT crdb_internal.payload_type(payload)`, then we would have to decode payload bytes (which we just encoded when inserting the row) every time we insert into the jobs table. 
___
*[`pkg/jobs/registry.go` line 517 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgr8Pm2lMDE0PYJFlf:-NIskKfiEtQsihN-VEfe:b7x372i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L517)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

wdyt about:

```
cols = []string{id, status, payload, progress, claim_session_id, claim_instance_id}
vals = []interface{}{jobID, StatusRunnnng, payloadBytes, progressBytes, s.ID().UnsafeBytes(), r.ID()}
placeholders := func() string {
  var p strings.Builder
  for i := range vals {
		if i > 0 {
			p.WriteByte(',')
		}
		p.WriteByte('$')
                p.WriteString(strconv.Itoa(i+1))
	}
	return p.String()
}
if r.settings.Version.IsActive(ctx, clusterversion.V23_1AddTypeColumnToJobsTable) {
   cols = append(cols, "type")
   vals = append(vals, jobType.String())
}
insertQuery = fmt.Sprintf(`INSERT INTO system.jobs (%s) VALUES (%s)`, strings.Join(cols, ","),  placeholders())

// Execute query ...
```
</blockquote></details>

Done. I like this. Again, I went with defaulting to more columns and doing `numCols -= 1` in the if condition.
___
*[`pkg/sql/sem/builtins/builtins.go` line 4044 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgxvYh08qUB0zYiuNI:-NIskXUB8C6TKaYkBLNK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/sql/sem/builtins/builtins.go#L4044)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you might be better off not using protoreflect package at all (which really exists so that you can create
protocol message of appropriate type, given type name).  You don't need it, because you know exactly the type of the message you expect.  So, something like this:

```
var msg jobspb.Payload
if  if err := protoutil.Unmarshal(data, &msg); err != nil {
   return nil, err...
}
return tree.NewDString(msg.Type().String()), nil
```

</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 4051 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIgw06R8-FbERocxoJL:-NIskZcwAfUNA5ae3u_D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/sql/sem/builtins/builtins.go#L4051)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

honestly, I'd just inline info, volatility, and return type here.  It's
pretty idiomatic in this file; also, why have variables when these are not
variables?
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIcfpQFIYvbyAi8tYYL-r1-55:-NIskacPDkHIHiihMPzk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type.go#L55)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

You could put these into a for loop so that you only need to write the code calling `migrateTable` once.
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type.go` line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIcfpQFIYvbyAi8tYYL-r1-70:-NIskbjt-xh5GJnP12gd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type.go#L70)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

```suggestion
	ie := d.InternalExecutorFactory.MakeInternalExecutorWithoutTxn()
	_, err := ie.Exec(ctx, "backfill-jobs-type-column", nil /* txn */, backfillTypeColumnStmt, username.RootUser)
```
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIglnmMA3Ofx_xVzXv4:-NIsknAk5ntMT-b3MBRH:b-fc6pns) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go#L93)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

is this safe to do?  Can this job start running (and likely crash)?
Or is it safe to do because the jobs system is not running yet? 
</blockquote></details>

It doesn't run. `CreateJobWithTxn` will not start the job. That's probably why this code doesn't cause issues. I'm okay with leaving this as is.

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andyyang890, @dt, and @miretskiy)

___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type.go` line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIsr8ow7Hgi1J6L0pEJ:-NIsr8ow7Hgi1J6L0pEK:b59dfoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7188761e414f8b1f56f88a88dc3c480a25bfb5/pkg/upgrade/upgrades/alter_jobs_add_job_type.go#L65)):*
> ```Go
> 
> 	ie := d.InternalExecutorFactory.MakeInternalExecutorWithoutTxn()
> 	_, err := ie.Exec(ctx, "backfill-jobs-type-column", nil /* txn */, backfillTypeColumnStmt, username.RootUser)
> ```

@andyyang890 Brought up a good point about putting both the backfill and schema change in the same migration. What happens in between the backfill ending and setting new cluster version? If we allow writes to the jobs table during this time (which I'm not sure we do), something could insert a job with a `NULL` `job_type` which would never be backfilled. Using two separate upgrades avoids this. cc @ajwerner @miretskiy

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 1 of 15 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andyyang890, @dt, and @jayshrivastava)

___
*[`pkg/jobs/registry.go` line 383 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIguSeD6mWpZb9OPbfv:-NJ0rb-h1MsRS-dpi5-f:b-7ql99t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/jobs/registry.go#L383)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Done. How do you feel about defaulting to `numColumns=8` and subtracting 1 if the version is old? The most common path is the one where the version is above the one with my upgrade. In the most common path, we can avoid the if branch.
</blockquote></details>

fine.
___
*[`pkg/jobs/registry.go` line 382 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NJ0rgMt6ZWWoKCFJeNZ:-NJ0rgMt6ZWWoKCFJeN_:bjc23pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bed397f45b4a9de61abf36b186cd84a761fd563/pkg/jobs/registry.go#L382)):*
> ```Go
> ) (string, []interface{}, []jobspb.JobID, error) {
> 	instanceID := r.ID()
> 	numColumns := 8
> ```

perhaps init columns, and then set numColumns := len(columns)
___
*[`pkg/jobs/registry.go` line 418 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NJ0rplF9vqy2iYmeKhd:-NJ0rplG2OrKjhEeAtNQ:b-pki3md) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bed397f45b4a9de61abf36b186cd84a761fd563/pkg/jobs/registry.go#L418)):*
> ```Go
> 
> 	// To run the upgrade below migration and schema change jobs will need
> 	// to be created using the old schema of the jobs table.
> ```

perhaps be specify and say ... "old schema, which does not have job_type column, ..."
___
*[`pkg/jobs/registry.go` line 499 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NJ0s2AI7MVhvYcNNcTt:-NJ0s2AI7MVhvYcNNcTu:btela5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bed397f45b4a9de61abf36b186cd84a761fd563/pkg/jobs/registry.go#L499)):*
> ```Go
> 	}
> 
> 	numCols := 7
> ```

dito  -- set numCols := len(cols)
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type.go` line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIsr8ow7Hgi1J6L0pEJ:-NJ0saF3-Sz0SkDvIWy_:b-4s0ifm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7188761e414f8b1f56f88a88dc3c480a25bfb5/pkg/upgrade/upgrades/alter_jobs_add_job_type.go#L65)):*
>  If we allow writes to the jobs table during this time (which I'm not sure we do), something could insert a job with a NULL job_type which would never be backfilled. Using two separate upgrades avoids this
hmmm... I don't think migration precludes another node from creating new jobs; Perhaps the code should assume that job_type can be null -- which, tbh, is not a big deal, I think.  Possible impact to some of the "show" jobs.
@ajwerner what say you?
___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIglnmMA3Ofx_xVzXv4:-NJ0stUwAkUP-y6OdYRJ:bik2pi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go#L93)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

It doesn't run. `CreateJobWithTxn` will not start the job. That's probably why this code doesn't cause issues. I'm okay with leaving this as is.
</blockquote></details>

Right, I understand CreateJobWithTxn does not start the job -- but that does not mean
that job registry is not running, and would not pick up this job immediately upon creation --
which will make this test flaky as it might crash the process.

You almost certainly want to get access to this node jobs registry, and set 	TestingResumerCreationKnobs 
for all job types to be your own resumer (that does nothing, or close to nothing).

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andyyang890, @dt, and @miretskiy)

___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIglnmMA3Ofx_xVzXv4:-NJ62dHNBBvqXhq1bF-Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e627b281b33cb04123d45b5eab55081702ac94a9/pkg/upgrade/upgrades/alter_jobs_add_job_type_test.go#L93)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Right, I understand CreateJobWithTxn does not start the job -- but that does not mean
that job registry is not running, and would not pick up this job immediately upon creation --
which will make this test flaky as it might crash the process.

You almost certainly want to get access to this node jobs registry, and set 	TestingResumerCreationKnobs 
for all job types to be your own resumer (that does nothing, or close to nothing).
</blockquote></details>

Done.

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andyyang890, @dt, and @miretskiy)

___
*[`pkg/upgrade/upgrades/alter_jobs_add_job_type.go` line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-NIsr8ow7Hgi1J6L0pEJ:-NJMUXYz-XEVG1jiWK4j:b9pz4hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a7188761e414f8b1f56f88a88dc3c480a25bfb5/pkg/upgrade/upgrades/alter_jobs_add_job_type.go#L65)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

>  If we allow writes to the jobs table during this time (which I'm not sure we do), something could insert a job with a NULL job_type which would never be backfilled. Using two separate upgrades avoids this
hmmm... I don't think migration precludes another node from creating new jobs; Perhaps the code should assume that job_type can be null -- which, tbh, is not a big deal, I think.  Possible impact to some of the "show" jobs.
@ajwerner what say you?
</blockquote></details>

We discussed this in person. I split the migration into two separate ones. This should be good to merge.

**Comment by miretskiy** [comment]
> We discussed this in person. I split the migration into two separate ones. This should be good to merge.

can you say more about the reasoning?

**Review by miretskiy — APPROVED** [review]

**Review by jayshrivastava — COMMENTED** [review]
Ofc! 

We have an upgrade that adds the type column to system.jobs and backfills it. The job registry code is updated to have a version gate: If cluster version is less than the new version, it will not write to the type column. Otherwise, write the type. 

The problem is that the version gate reads the cluster version. The cluster version is updated some period of time after the upgrade. During this period of time, we may write jobs to the jobs table without the type column populated. Because this happens after the backfill, the type columns will remain NULL for these rows.

The solution was to break this up into two upgrades / versions. The first adds the nullable type column. The registry will populate the type column if the cluster version is greater or equal to this version. Then, we do the backfill upgrade. This leaves no gaps for rows to sneak into the table without a populated type.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93144#-:-NJMkMTa0rXi2o2NEL9u:b-mkyme0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andyyang890, @dt, and @miretskiy)

**Comment by rickystewart** [comment]
bors r-

Breaking CI
